onecodex/models/collection.py

```diff
@@ -389,18 +392,181 @@ def taxonomy(self):
+        metric : str
+            Metric values to return,
+            `{'coverage', 'abundance'}` for `annotation==onecodex.lib.enum.FunctionalAnnotation.Pathways`,
+            `{'rpk', 'cpm'}` for other annotations
+        fill_missing : bool, optional
+            Fill `np.nan` values
+        filler : float, optional
+            Value with which to fill `np.nan` values
+        """
+        # validate args
+        annotation = FunctionalAnnotations(annotation)
+        if annotation == FunctionalAnnotations.Pathways:
+            if metric not in ["coverage", "abundance"]:
+                raise ValueError(
+                    "if using annotation='pathways', 'value' must be one of ['coverage', 'abundance']"
+                )
+        elif metric not in ["cpm", "rpk"]:
+            raise ValueError(
+                f"if using annotation={annotation.value}, 'value' must be one of ['cpm', 'rpk']"
+            )
+
+        data = {}
+        all_features = set()
+        # iterate over functional profiles, subset data, and store in data dict
+        for profile in self._functional_profiles:
+            # get table using One Codex API
+            table = profile.table(annotation=annotation, taxa_stratified=taxa_stratified)
+            # filter by indicated metric
+            if not taxa_stratified:
+                table = table[table["metric"] == "total_" + metric]
+            else:
+                table = table[table["metric"] == metric]
+            # if taxa stratified, concatenate id and taxon_name
+            if taxa_stratified:
+                table["id"] = table["id"] + "_" + table["taxon_name"]
+            # store tables for later retrieval
+            data[profile.id] = dict(zip(table.id, table.value))
+            all_features.update(set(table["id"]))
+
+        features_to_ix = {}
+        feature_list = []
+        for ix, feature in enumerate(all_features):
+            features_to_ix[feature] = ix
+            feature_list.append(feature)
+
+        # initialize an array and fill it
+        array = np.full(shape=(len(data), len(features_to_ix)), fill_value=np.nan)
+        profile_ids = []
+        for profile_index, profile_id in enumerate(data):
+            for feature_id in data[profile_id]:
+                array[profile_index, features_to_ix[feature_id]] = data[profile_id][feature_id]
+            profile_ids.append(profile_id)
+        df = pd.DataFrame(
+            array, index=pd.Index(profile_ids, name="functional_profile_id"), columns=feature_list
+        )
+
+        if fill_missing:
+            df.fillna(filler, inplace=True)
+        self._cached["functional_results"] = df
+        self._cached["functional_results_content"] = {
+            "annotation": annotation,
+            "taxa_stratified": taxa_stratified,
+            "metric": metric,
+            "fill_missing": fill_missing,
+            "filler": filler,
+        }
+
+    def _functional_results(self, **kwargs):
```

Comment:
Here's how I'd suggest handling the kwarg matching:

1. use this method signature (i.e. make all parameters *required positional arguments*):
    ```python
    def _collate_functional_results(self, annotation, metric, taxa_stratified, fill_missing, filler):
        ...
    ```
2. `_functional_results()` becomes:
    ```python
    def _functional_results(self, **kwargs):
        if "functional_results" not in self._cached:
            self._collate_functional_results(**kwargs)
            return self._cached["functional_results"]

        if kwargs != self._cached["functional_results_content"]:
            self._collate_functional_results(**kwargs)

        return self._cached["functional_results"]
    ```

The benefits of this approach are:
 - we only define the default values once (in `_to_functional_df()`)
 - if a new parameter gets added, it only has to be added to `_to_functional_df()` signature and to `_collate_functional_results()`
 - if we *forget* to pass a parameter through to `_collate_functional_results()`, it'll error because all arguments are required

Reply:
This significantly simplifies things. I definitely got wrapped up in my own historical spaghetti as I developed these methods, thanks for helping me see past it!